### ISSUE TITLE
clarify the duty of `common-store-api` and `common-store-api-sdk`

### ISSUE THREAD

**Issue by drmingdrmer** [datafuselabs/databend#1940]
It seems like `store-api` defines the abstract data types and interfaces, while `store-api-sdk` provides concrete implementations.

If it is true, `store-api` should let one access an abstract KVApi without depending on `store-api-sdk`.

E.g. `UpsertKVAction` is defined in `store-api-sdk` but the corresponding `UpsertKVActionResult` is defined in `store-api`.

https://github.com/datafuselabs/databend/blob/f8c569d9a29f509bede5756b3a1309e65fe40287/common/store-api-sdk/src/impl_flights/kv_api_impl.rs#L133-L139

https://github.com/datafuselabs/databend/blob/f8c569d9a29f509bede5756b3a1309e65fe40287/common/store-api/src/kv_apis/kv_api.rs#L22-L28

**Comment by drmingdrmer** [comment]
@dantengsky @BohuTANG @ariesdevil 

By moving implementation irrelevant types into `store-api` would reduce some potential unexpected dependency cycles.

**Comment by dantengsky** [comment]
> It seems like `store-api` defines the abstract data types and interfaces, while `store-api-sdk` provides concrete implementations.
> 

yes

> If it is true, `store-api` should let one access an abstract KVApi without depending on `store-api-sdk`.

`store-api` is for crates like `common-management` which do not depends on the impl of API.


please feel free to refactor them

**Comment by dantengsky** [comment]
> @dantengsky @BohuTANG @ariesdevil
> 
> By moving implementation irrelevant types into `store-api` would reduce some potential unexpected dependency cycles.

currently, the implementation of `KVApi::upsert_kv` wraps the args in a `UpsertKVAction`, since the signature of `do_action` requires that.

```
impl KVApi for StoreClient {
    #[tracing::instrument(level = "debug", skip(self, value))]
    async fn upsert_kv(
        &self,
        key: &str,
        seq: MatchSeq,
        value: Option<Vec<u8>>,
        value_meta: Option<KVMeta>,
    ) -> Result<UpsertKVActionResult> {
        self.do_action(UpsertKVAction {
            key: key.to_string(),
            seq,
            value,
            value_meta,
        })
        .await
    }
```

should we expose `UpsertKVAction` to the API user directly?, and change the signature of `KVApi::upsert_kv` to 

```
async fn upsert_kv(
        &self,
        action: UpsertKVAction
    ) -> Result<UpsertKVActionResult> 
```

I am not quite sure, I think the current signature is more natural. Or did I missed anything?

**Comment by dantengsky** [comment]
btw, Server side impls of Api,  are supposed to  depend `store-api-sdk` only,   for types originally defined in `store-api` which are not exported by `store-api-sdk`,  they are "invisible" to the server side.  The `ActionHandler`s in crate `store` follows this rule or at least used to be following.

**Comment by drmingdrmer** [comment]
> should we expose `UpsertKVAction` to the API user directly?, and change the signature of `KVApi::upsert_kv` to

Structured input/output types would always be better. Anything is easy to change is better.
A struct would be more friendly when searching for usage, adjusting argument order, adding/removing fields etc.
And when refactoring codes, one only needs to update the caller and the final receiver, not the entire calling chain.

**Comment by drmingdrmer** [comment]
> btw, Server side impls of Api, are supposed to depend `store-api-sdk` only, for types originally defined in `store-api` which are not exported by `store-api-sdk`, they are "invisible" to the server side. The `ActionHandler`s in crate `store` follows this rule or at least used to be following.

I'm afraid `store` can not be implemented without `UpsertKVActionResult` which is defined in `store-api`.

It does not smell weird if a server has to depend on some abstract data types or traits, but it smells weird that it has to depend on a crate that impls the client to the server.

**Comment by dantengsky** [comment]
Hi bro, my thoughts on the above issues, hope it may be useful:

> Structured input/output types would always be better. Anything is easy to change is better.
> A struct would be more friendly when searching for usage, adjusting argument order, adding/removing fields etc.
> And when refactoring codes, one only needs to update the caller and the final receiver, not the entire calling chain.

`store-api` is for service users, they are built upon the backend services, we may provide apis that are quite different from the service exposed by the backend server. take KVApi for example,  we may provide an unconditional `delete` method by using backend `upsert` service api, or wrap the `upsert_kv` method so that the API users do not need to check the (prev, result) tuple to deduce the result... without changing the backend impls (and the interface they expose).

I really do not know, among ergonomic api and ease of refactoring, which is more important, imho,  for strongly typed lang like rust, refactoring is less risky (but not riskless of course).


> I'm afraid store can not be implemented without UpsertKVActionResult which is defined in store-api.
> 
> It does not smell weird if a server has to depend on some abstract data types or traits, but it smells weird that it has to depend on a crate that impls the client to the server.

I agree with you that it may be necessary to extract some common types that used by both store and sdk to a separated crate, if I get it right.   since the SDK is in rust (the same lang that backend uses), it is supposed to provide some common types (which could be re-used by backend directly). but I do not strongly adhere to this, it is not unusual to convert the message to totally different types in backend while de/serializing the req/resp messages (e.g. the extreme case is SDK is not implemented in rust)

but I think this is not the root cause of the "dependency cycles" issue. rather, I think, 
 - the crate `metasrv` need to be split into more fine-grained crates.
 - `sm.rs` should NOT depend on types defined in store-api-sdk, as the `LocalKVStore` uses `StateMachine`, which will lead to dependency troubles (and the `ActionHandlers` may need to be extracted as well)
 -  other components of data-bend are not suggested to be dependent on crate `metasrv`,  usually, no libs or bin are supposed to be depend on another bin
     Of course, there are always exceptions, if I recall correctly, CH's Server and Client are the same binary. 

did you decide where the `LocalKVStore` is being moved to? maybe I can help.

Cheers bro, happy coding

**Comment by drmingdrmer** [comment]
> I really do not know, among ergonomic api and ease of refactoring, which is more important, imho, for strongly typed lang like rust, refactoring is less risky (but not riskless of course).

It's not about risk. It's about lines of code you have to change to get one refactoring task done.

**Comment by drmingdrmer** [comment]
> did you decide where the `LocalKVStore` is being moved to? maybe I can help.

The first that needs to be done is to extract `StateMachine` off `metasrv`. Then localkv would find its place easily. Likely to be a `common` crate.

Some refactoring work that needs to be done is to create a standalone `KVClient` out of `StoreClient` etc. These may block future works in next couple of weeks.

**Comment by dantengsky** [comment]
> Some refactoring work that needs to be done is to create a standalone KVClient out of StoreClient etc. These may block future works in next couple of weeks.

does it mean that we should provide a dedicated crate, which provides `KVClient` only?

**Comment by BohuTANG** [comment]
This may be a better idea.
One is KVClient, it's for the metaservice.
Another is DFSClient? It's for dfs store.
We mixed them all in one, a bit difficult to understand :)

**Comment by drmingdrmer** [comment]
> This may be a better idea.
> One is KVClient, it's for the metaservice.
> Another is DFSClient? It's for dfs store.
> We mixed them all in one, a bit difficult to understand :)

That's exactly what I want!

**Comment by drmingdrmer** [comment]
> does it mean that we should provide a dedicated crate, which provides `KVClient` only?

Very likely to be. Since the `StoreClient` is provided in crate `store-api-sdk`, as I recalled, having a kvclient in it might confuse readers.

**Comment by dantengsky** [comment]
> This may be a better idea.
> One is KVClient, it's for the metaservice.
> Another is DFSClient? It's for dfs store.
> We mixed them all in one, a bit difficult to understand :)

Yes, the current store-api/sdk reflects the former design, they are being splitter in PR #1952

**Comment by dantengsky** [comment]
> Very likely to be. Since the `StoreClient` is provided in crate `store-api-sdk`, as I recalled, having a kvclient in it might confuse readers.

yeah, as we finally decided recently to split the services,  combine all the apis together no longer make sense. especially, crate like `management` emerges, which depends KVApi only.

**Comment by dantengsky** [comment]
> Very likely to be. Since the StoreClient is provided in crate store-api-sdk, as I recalled, having a kvclient in it might confuse readers.

:D, I misunderstood your words. as far as I can tell, it is more common to provide client in package SDK rather in package API. correct me if I mistaken

**Comment by dantengsky** [comment]
as pr #1960 has been merged, "dependency cycles" gone away, shall we close this issue? cc @drmingdrmer

**Comment by drmingdrmer** [comment]
> as pr #1960 has been merged, "dependency cycles" gone away, shall we close this issue? cc @drmingdrmer

One still needs both `store-api`(xxxActionResult is defined here) and `store-api-sdk`(xxxAction is defined here) to accomplish a job.

**Comment by dantengsky** [comment]
> One still needs both `store-api`(xxxActionResult is defined here) and `store-api-sdk`(xxxAction is defined here) to accomplish a job.

which one? `common-management` should not depend on `store-api-sdk`

**Comment by dantengsky** [comment]
bro, I think the duties of api and sdk are not that hard to be recognized, no offense, but it is a common naming convention.

currently:

`common-management` depends `common-store-api` only
`kvlocal` depends `common-store-api` only
`metasrv` depends `common-store-api-sdk` only
`store` depends `common-store-api-sdk` only

and "dependency cycles" no longer exist (I "feel", it is the main motivation of this issue)

pls allow me to close this issue, the atmosphere here is not that happy, let's meet up offline, "nothing can not be resolved with a BBQ and beers"

**Comment by drmingdrmer** [comment]
`store-api` does not provide complete API for a crate to use.
E.g., the crate `raft-store`, which only needs some interface types, has to depend on `store-api-skd`.

Every crate depends on `store-api` has to depend on `store-api-sdk`.

![image](https://user-images.githubusercontent.com/44069/134758916-bc0bb6cd-f264-4459-9ba6-da13ab5375b2.png)


```
cargo tree -i common-store-api

common-store-api
├── common-management
│   └── databend-query
│       ├── databend-cli
│       └── fuzz
├── common-store-api-sdk
│   ├── common-raft-store
│   │   ├── databend-store
│   │   │   └── databend-cli
│   │   ├── kvlocal
│   │   │   ├── common-management
│   │   │   └── databend-query
│   │   └── metasrv
│   │       └── databend-store
│   ├── databend-query
│   ├── databend-store
│   └── metasrv
├── databend-query
└── kvlocal
```

**Comment by dantengsky** [comment]
> Every crate depends on store-api has to depend on store-api-sdk

No, that is not the case. pls check ... crate `common-management`?
`xxx-api`s  are for crates like it.

if the dependency relations should be tweaked, shall we open another issue?

**Comment by drmingdrmer** [comment]
`common-management` has to indirectly depends on `store-api-sdk` because of one of its dependency.

And it is not only about `common-management`.  It is about any crate that needs just the store API but does not care about the impl. 

In this case, it is `raft-store`. `raft-store` does not care about how a `store` is implemented, thus it should not depend on `store-api-sdk`.

**Comment by dantengsky** [comment]
> common-management has to indirectly depends on store-api-sdk because of one of its dependency.
may i ask which one? I do not have a laptop at hand now

**Comment by dantengsky** [comment]
> > common-management has to indirectly depends on store-api-sdk because of one of its dependency.
> may i ask which one? I do not have a laptop at hand now



> > common-management has to indirectly depends on store-api-sdk because of one of its dependency.

find it,  it is `kvlocal`.  
`common-management` need not depend on `kvlocal` except for `dev-dependencies`, seems you are working on it.

`raft-store` seems to be able to migrate to depend on API only, let me try it

**Comment by drmingdrmer** [comment]
> > common-management has to indirectly depends on store-api-sdk because of one of its dependency.

> may i ask which one? I do not have a laptop at hand now

It seems somehow you missed the deps graph I pasted up there in a previous comment. Put it here:

`management` indirectly depends on `raft-store` and `raft-store` needs data types from `store-api-sdk`.

```
cargo tree -i common-store-api

common-store-api
├── common-management
│   └── databend-query
│       ├── databend-cli
│       └── fuzz
├── common-store-api-sdk
│   ├── common-raft-store
│   │   ├── databend-store
│   │   │   └── databend-cli
│   │   ├── kvlocal
│   │   │   ├── common-management
│   │   │   └── databend-query
│   │   └── metasrv
│   │       └── databend-store
│   ├── databend-query
│   ├── databend-store
│   └── metasrv
├── databend-query
└── kvlocal
```

![](https://user-images.githubusercontent.com/44069/134758916-bc0bb6cd-f264-4459-9ba6-da13ab5375b2.png)

**Comment by drmingdrmer** [comment]
> `common-management` need not depend on `kvlocal` except for `dev-dependencies`, seems you are working on it.

I am not working on it yet. But I'd like to have it done right now.

**Comment by drmingdrmer** [comment]
- 👍 👍 👍  `raft-store` no longer depends on `xxx-sdk`. Fixed in #1967 Thanks to @dantengsky !

**Comment by drmingdrmer** [comment]
> > > common-management has to indirectly depends on store-api-sdk because of one of its dependency.
> > > may i ask which one? I do not have a laptop at hand now
> 
> > > common-management has to indirectly depends on store-api-sdk because of one of its dependency.
> 
> find it, it is `kvlocal`.
> `common-management` need not depend on `kvlocal` except for `dev-dependencies`, seems you are working on it.
> 
> `raft-store` seems to be able to migrate to depend on API only, let me try it

- fixed in #1973 


All problems in this issue seem to be solved? 🤔

**Comment by BohuTANG** [comment]
Let's close it :)
